src/Servers/IIS/AspNetCoreModuleV2/RequestHandlerLib/filewatcher.cpp

```diff
@@ -261,16 +297,94 @@ HRESULT
         }
     }
 
-    if (fFileChanged && !_lStopMonitorCalled)
+    if (fAppOfflineChanged && !_lStopMonitorCalled)
     {
         // Reference application before
         _pApplication->ReferenceApplication();
         RETURN_LAST_ERROR_IF(!QueueUserWorkItem(RunNotificationCallback, _pApplication.get(), WT_EXECUTEDEFAULT));
     }
 
+    if (fDllChanged && m_fShadowCopyEnabled && !_lStopMonitorCalled)
+    {
+        // Reset timer for dll checks
+        LOG_INFO(L"Detected dll change, resetting timer callback which will eventually trigger shutdown.");
+        m_Timer.CancelTimer();
+        m_Timer.InitializeTimer(FILE_WATCHER::TimerCallback, this, 5000, 3000);
+    }
+
     return S_OK;
 }
 
+
+VOID
+CALLBACK
+FILE_WATCHER::TimerCallback(
+    _In_ PTP_CALLBACK_INSTANCE Instance,
+    _In_ PVOID Context,
+    _In_ PTP_TIMER Timer
+)
+{
+    Instance;
+    Timer;
+    CopyAndShutdown((FILE_WATCHER*)Context);
+}
+
+DWORD WINAPI FILE_WATCHER::CopyAndShutdown(LPVOID arg)
+{
+    auto watcher = (FILE_WATCHER*)arg;
+
+    // Only copy and shutdown once
+    SRWExclusiveLock lock(watcher->m_copyLock);
+    if (watcher->m_copied)
+    {
+        return 0;
+    }
+
+    watcher->m_copied = true;
+
+    LOG_INFO(L"Starting copy on shutdown in filewatcher, creating directory.");
+
+    auto directoryNameInt = 0;
+    auto currentShadowCopyDirectory = std::filesystem::path(watcher->m_shadowCopyPath);
+    auto parentDirectory = currentShadowCopyDirectory.parent_path();
+    try
+    {
+        directoryNameInt = std::stoi(currentShadowCopyDirectory.filename().string());
+    }
+    catch (...)
+    {
+        OBSERVE_CAUGHT_EXCEPTION();
```

Comment:
Does this macro log somewhere?

Reply:
Yeah to the debug log. I don't think this should fail though; it's mostly to avoid crashing the process.